edb/edgeql/compiler/inference/cardinality.py

```diff
@@ -284,32 +322,248 @@ def _find_visible(
         else:
             path_id = ir.path_id
 
+        # This is a nasty hack, I think. If a path is an unfenced
+        # descendant, don't find a visible node above us. This fixes
+        # issues like in test_edgeql_scope_filter_01 where we mean to
+        # refer to our ought-to-be-namespaced child but instead find
+        # something else.
```

Comment:
Or maybe we should just xfail test_edgeql_scope_filter_01?

Reply:
Hm. Yea, the `strip_weak_namespaces` business is all wrong.  What happens if you run cardinality inference _before_ `strip_weak_namespaces`?

Reply:
Yeah, that works!

I had actually tried that earlier, and it didn't work, but when I went back so I could explain what didn't work about it, and it turns out I had moved cardinality inference to happen in between the two parts of stripping weak namespaces, which unsurprisingly works less well than doing it either before or after.